### DISCUSSION TITLE
Alias/let syntax interpretation and use

### DISCUSSION THREAD

**Discussion by extemporalgenome** [cuelang/cue#687] [Q&A] [answered]
In the CUE language spec in the Aliases section (https://cuelang.org/docs/references/spec/#aliases) is the following syntactic form:

```
foo: X  // 4
X="not an identifier": 4
```

What is the purpose/meaning of the `"not an identifier"` part (it doesn't appear in the example output)?

Is it just a throwaway expression to syntactically distinguish normal fields from aliases when used in this way?

How would this be different from `let` declarations (which the tutorial describes as aliases, while the spec description of `let` does not use the term alias).

How are aliases and `let` declarations distinct from `_#` definitions?

Thanks
- Kevin

**Comment by mpvl** [comment, marked as answer]
The purpose of the `X` in `X="not an identifier": 4` is that it allows referencing the field. The field has a name that is not a valid identifier, so there would otherwise be no way to directly reference the field. For instance, the `foo: X` references that field and will evaluate to `foo: "not an identifier": 4`.

A `let` introduces a name for an expression that can be reused without it being part of the output.
So, where:
```
X=a: 4
b: X
```
would evaluate to 
```
a: 4
b: 4
```
using a let
```
let X = 4
b: 4
```
evaluates to
```
b: 4
```
instead.

In general if you see a `X = ` it introduces a new name for the element it precedes, without otherwise changing the meaning of this element. It can be used to work around a host of issues such as referring to fields that are otherwise invalid names, or just long, generated field names (such as in X="key\(x)": int`) or elements that are otherwise hard or impossible to reference.

A `let` on the other hand, is used to hoist an often-repeated expression or just to break an expression into smaller chunks, without having to introduce a new field for that, for instance:
```
let byte = "[0-9a-x]{2}"
#hexIP: =~"\(byte).\(byte).\(byte).\(byte)"
```

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/687.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
